cub/cub/warp/specializations/warp_reduce_shfl.cuh

```diff
@@ -43,29 +43,60 @@ CUB_NAMESPACE_BEGIN
 
 namespace detail
 {
-template <class A = int, class = A>
-struct reduce_add_exists : ::cuda::std::false_type
-{};
+template <class T, class ReductionOp, class = void>
+inline constexpr bool reduce_add_exists = false;
 
 template <class T>
-struct reduce_add_exists<T, decltype(__reduce_add_sync(0xFFFFFFFF, T{}))> : ::cuda::std::true_type
-{};
+inline constexpr bool reduce_add_exists<T, ::cuda::std::plus<>, decltype(__reduce_add_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
 
-template <class T = int, class = T>
-struct reduce_min_exists : ::cuda::std::false_type
-{};
+template <class T>
+inline constexpr bool reduce_add_exists<T, ::cuda::std::plus<T>, decltype(__reduce_add_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
+
+template <class T, class ReductionOp, class = void>
+inline constexpr bool reduce_min_exists = false;
+
+template <class T>
+inline constexpr bool reduce_min_exists<T, ::cuda::minimum<>, decltype(__reduce_min_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
+
+template <class T>
+inline constexpr bool reduce_min_exists<T, ::cuda::minimum<T>, decltype(__reduce_min_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
+
+template <class T, class ReductionOp, class = void>
+inline constexpr bool reduce_max_exists = false;
+
+template <class T>
+inline constexpr bool reduce_max_exists<T, ::cuda::maximum<>, decltype(__reduce_max_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
+
+template <class T>
+inline constexpr bool reduce_max_exists<T, ::cuda::maximum<T>, decltype(__reduce_max_sync(0xFFFFFFFF, T{}))> =
+  (::cuda::std::is_same_v<int, T> || ::cuda::std::is_same_v<unsigned int, T>);
```

Comment:
Note to self:

I believe we can relax that constrain to `is_integral<T> && sizeof(T) <= sizeof(int)` and then in the optimization  pass below cast to `using promoted_t = ::cuda::std::conditional_t<::cuda::std::is_signed_v<T>, int, unsigned>;`

Then casting back to the original T should be fine. That said, I believe this to be a rather invasiev change so I leave that for a followup